SeqAnalysis.pm. . . . . Perl module containing often-used subroutines for sequence processing<br>
bio3d.R . . . . . . . . Loads Bio3D module and runs through steps to generate PCA plot<br>
bp_sreformat.pl . . . . <br>
bp_ssearchparse.pl. . . Parses all ssearch files in the given directory; Gathers counts and sequences.<br>
dephylip_tree.pl. . . . Dephylips Newick tree files<br>
fasta2hmm.pl. . . . . . Pipeline for converting unaligned fasta files to hmm files<br>
getkegg.pl. . . . . . . Parses a keggfile and downloads the genes <br>
getscripts.pl . . . . . Produces README.md containing all of the custom scripts in ~/scripts<br>
gimpvert.pl . . . . . . Converts gel-imager .tif images to .png files; opens in Gimp for optional editing<br>
hmm_run.pl. . . . . . . Generates a shell script to run batch searches using either hmmsearch or hmmscan<br>
hmmparse.pl . . . . . . Parses an HMM result file, scan or search<br>
kegg.pl . . . . . . . . Queries the KEGG db to collect genes in a given pathway<br>
mkpl.pl . . . . . . . . Sets up perl skeleton script<br>
mkpm.pl . . . . . . . . Sets up a perl module template with comments and standard info<br>
pf2g.pl . . . . . . . . Takes a list of pfam IDs and maps them to go terms<br>
print_all.R . . . . . . Loads Bio3D module and plots PCA using a variety of pre-defined color schemes<br>
process_kegg.pl . . . . Process a KEGG record file<br>
showPool.pl . . . . . . Takes pool input and highlights residues in Pymol<br>
spc2us.pl . . . . . . . renamer script used for replacing spaces w/ underscores in filenames<br>
ssearch_run.pl. . . . . Generates batch shell script for ssearches<br>
ssearchrank.pl. . . . . Searches many FASTA search result files to get a better scoring match for a particular transcript<br>
