SeqAnalysis.pm. . . . . Perl module containing often-used subroutines for sequence processing<br>
bp_sreformat.pl . . . . <br>
bp_ssearchparse.pl. . . Parses all ssearch files in the given directory; Gathers counts and sequences.<br>
convert_all.pl. . . . . Converts all files of one type in a directory to a second type using bp_sreformat<br>
dephylip_tree.pl. . . . Dephylips Newick tree files<br>
gbk_extract.pl. . . . . Pulls out descriptions from genbank files<br>
getscripts.pl . . . . . Produces README.md containing all of the custom scripts in ~/scripts<br>
gimpvert.pl . . . . . . Converts gel-imager .tif images to .png files; opens in Gimp for optional editing<br>
hmm_run.pl. . . . . . . Generates a shell script to run batch searches using either hmmsearch or hmmscan<br>
hmmparse.pl . . . . . . Parses an HMM result file, scan or search<br>
kegg.pl . . . . . . . . Queries the KEGG db to collect genes in a given pathway<br>
mkpl.pl . . . . . . . . Sets up perl skeleton script<br>
mkpm.pl . . . . . . . . Sets up a perl module template with comments and standard info<br>
modeller_loop.py. . . . Homology modeling by the automodel class<br>
motif.pl. . . . . . . . <br>
pf2g.pl . . . . . . . . Takes a list of pfam IDs and maps them to go terms<br>
seqlen.pl . . . . . . . Extracts sequence lengths from a fasta file<br>
showPool.pl . . . . . . Takes pool input and highlights residues in Pymol<br>
spc2us.pl . . . . . . . renamer script used for replacing spaces w/ underscores in filenames<br>
split_fastq.pl. . . . . Splits a fastq file into a fasta and a qual file<br>
sra_to_fnaqual.pl . . . Uses fastq-dump and split_fastq.pl to convert an .sra file to the corresponding .fna and .qual files<br>
ssearch_run.pl. . . . . Generates batch shell script for ssearches<br>
ssearchrank.pl. . . . . Searches many FASTA search result files to get a better scoring match for a particular transcript<br>
